SPI: support for ISR and work_queue driven transfers

dakejahl opened the pull request:
Added an `_is_locked` variable to the `SPI` class which is a simple bit mask where the bit position corresponds to the SPI bus number. When running `SPI::transfer` from an ISR, `LockMode` is selected as `LOCK_NONE` (because you cannot pend on a sem in interrupt context). I have added a check for the `_is_locked` flag for the given bus when `LockMode` is none, thus preventing an ISR driven `SPI::transfer` from clobbering an already active transfer taking place in workqueue/task/etc.

dakejahl commented:
Next steps, test this with two IMUs (6500/9250) running on the same SPI bus (SPI1). Schedule one using the new `px4_workq` and the other with the current `hrt_call_every`. Look at `spi_isr_deferred` perf counter to see if the ISR driven transfer is being occasionally aborted.

dakejahl commented:
Merged into #11261 and tested like so
```
	if(_whoami == MPU_WHOAMI_6500) {
		ScheduleOnInterval(_call_interval - MPU9250_TIMER_REDUCTION, 10000);
	} else {

		hrt_call_every(&_call,
			       1000,
			       _call_interval - MPU9250_TIMER_REDUCTION,
			       (hrt_callout)&MPU9250::measure_trampoline, this);
	}
```
This appears to be working correctly
![selection_018](https://user-images.githubusercontent.com/37091262/51772898-73275f00-20aa-11e9-8be9-b875143d8e34.png)

And I am seeing perf counts for deferred ISR driven transfers
![selection_017](https://user-images.githubusercontent.com/37091262/51772909-7b7f9a00-20aa-11e9-907c-8a02a378ec9c.png)

@dagar

dakejahl commented on src/lib/drivers/device/nuttx/SPI.cpp:
```diff
@@ -84,87 +83,99 @@ SPI::SPI(const char *name,
+
+void SPI::lock(struct spi_dev_s *dev)
+{
+	_is_locked |= 1 << (_device_id.devid_s.bus - 1);
```
I don't think so, as long as the member variable is set first and unset last an ISR will defer. But could make it atomic for good measure.

Okay I'll take a look at the other platforms bus numbering.

dakejahl replied:
A search inside of `Firmware/boards` for **PX4_SPI_BUS** and I am seeing all 1 based numbering.

davids5 replied:
Have a look at https://github.com/PX4/Firmware/blob/master_fmuv4_runtime_spi4/src/platforms/px4_micro_hal.h#L50-L51 - the key is to see that is in `_device_id.devid_s.bus' on a Kinetis - me thinks it is 0. To make this safe I would use an uint_32 here and a bit 0 would be bus0

dakejahl replied:
Okay easy enough :smile:

dakejahl commented on src/lib/drivers/device/nuttx/SPI.cpp:
```diff
@@ -84,87 +83,99 @@ SPI::SPI(const char *name,
-
 	if ((send == nullptr) && (recv == nullptr)) {
-		return -EINVAL;
+		return PX4_ERROR;
```
Yeah I know I just wasn't sure if it mattered because we are never checking the return value. I'll change it back.

dakejahl commented:
I removed result because `_transfer` was always returning `OK`. And yeah nuttx SPI `exchange` has no return value.

dakejahl commented:
I need to do one more test on this Monday, I had left out applying the bus number to the bitmask when checking if the bus was locked.

dakejahl commented:
Okay I think this is done now. I fixed an issue where I wasn't checking the bit position in the `_is_locked` bit mask. So @dagar the ISR driven cycle is actually deferring **much less** than what was apparent previously (previously I was just looking at if _anyone_ was using _any_ SPI bus, i.e activity on SPI2). 
![selection_022](https://user-images.githubusercontent.com/37091262/51853881-9dfdf700-22e6-11e9-96b0-a9978a6880b0.png)


This actually isn't relevant though because the default `LOCK_MODE` is `LOCK_PREEMPTION` anyways (enters critical section for `transfer`). However I changed it to `LOCK_THREADS` for the purpose of demonstrating this `_is_locked` mechanism is functional.

Bench tested on a Teal with 2 IMUs(6500/9250) on SPI1, one running from the new `wq` and the other from an `hrt_call_every`

bkueng commented on src/lib/drivers/device/nuttx/SPI.cpp:
```diff
@@ -84,87 +78,99 @@ SPI::SPI(const char *name,
+
+void SPI::lock(struct spi_dev_s *dev)
+{
+	_is_locked |= (1 << _device_id.devid_s.bus);
```
@dakejahl this needs to be atomic. It's a shared state potentially written to by multiple threads.
You can use the API in https://github.com/PX4/Firmware/pull/11328.

Also this assumes that all `SPI` instances for a given bus run on the same thread. Can you add this as comment here to make the assumption explicit?

dagar replied:
FYI once we're sure there's no longer a mix of PX4 SPI drivers running out of HRT and threads we can drop all of this and lean on the underlying nuttx SPI_LOCK (semaphore).

dagar replied:
What I meant to emphasize was that this only exists for the case of a potential HRT and thread SPI conflict.

dakejahl replied:
> this needs to be atomic. It's a shared state potentially written to by multiple threads.
You can use the API in #11328.

Cool! I'll add that in as soon as #11328 is merged.

dakejahl commented:
> Also this assumes that all SPI instances for a given bus run on the same thread.

Could you explain please?

bkueng commented:
>  Could you explain please? I don't think it does.

The implementation is not correct if this does not hold. This is what can go wrong:
- 2 threads enter `SPI::lock` with the same bus id.
- both set `_is_locked`
- one of the threads (let's say thread1) enters the critical section, the other one (thread2) blocks
- thread1 leaves the CS and clears the bit in _is_locked.
- thread2 enters the CS, while the bit in _is_locked is cleared.

dakejahl commented:
@bkueng Thanks for the explanation! I've updated the code. Lock free atomic operations are a new concept to me, please let me know if this doesn't look right.

davids5 commented:
@dakejahl @dagar - Is it true that in the end (all done), there is only 1 thread per bus and the possibility of hrt interruption.?

If so:There is a 1:1 of bus:semaphore so this reduce to  

`interrupt context && dev->semaphore count  == 0 -> exit`   

So the loser is always running in interrupt context. Therefore the test does not need protection at (it is non interruptible - no nesting or HI priority IRQ) So `dev->semaphore count  == 0 => Exit.`  the sem_wait is already wrapped in a DI, EI and is safe.

Is the problem a shared bus? Barro and Nuttx driver on same bus?

dagar commented:
I believe the only reason that wasn't done is accessing the somewhat hidden spidev semaphore.

dakejahl commented:
Yeah I wasn't sure how to get at that semaphore.

bkueng commented on src/lib/drivers/device/nuttx/SPI.cpp:
```diff
@@ -84,94 +79,108 @@ SPI::SPI(const char *name,
+{
+	SPI_LOCK(dev, false);
+	// _is_locked &= ~(1 << _device_id.devid_s.bus);
+	_is_locked.fetch_nand(1 << _device_id.devid_s.bus);
```
```suggestion
	_is_locked.fetch_and(~(1 << _device_id.devid_s.bus));
```

dakejahl replied:
thanks

dakejahl commented:
> Given the " at-most-one-thread-per-bus assumption" has a complication we need to resolve to make it so.

I think this makes sense architecturally and we should enforce it and document it.
> We should also change the parameter to not be written during flight. 

You are referring to `_locking_mode`? I agree. 

> That will fully solve the barro noise issue. If we had adequate holdup time and a power fail detection we could commit them with an orderly shutdown as can be done with SW reboot. But that is a whole other issue.

I think the way the ms5611 collects is:
1. Schedules the measurement on the `hp_wq` -> `spi::transfer` (tells ms5611 to measure)
2. Reschedules the collection phase on the `hp_wq` for some time in the future (meanwhile ms5611 is performing a measurement, the SPI bus is **not** locked, and param module can use the bus

I think we would need to keep the `SPIx` bus locked for the entirety of the measure/collect of the ms5611 if the condition is "don't wiggle SPI while doing a conversion"

Yeah?

davids5 commented:
>>We should also change the parameter to not be written during flight.

>You are referring to _locking_mode? I agree

No param save to the mtd.

>I think the way the ms5611 collects is:

>Schedules the measurement on the hp_wq -> spi::transfer (tells ms5611 to measure)
Reschedules the collection phase on the hp_wq for some time in the future. Meanwhile ms5611 is performing a measurement, the SPI bus is not locked, and param module can use the bus.

>>I think we would need to keep the SPIx (spi2 on fmu-v4) bus locked for the entirety of the measure/collect of the ms5611 if the condition is "don't wiggle SPI while doing a conversion".

I agree but only on HW where it matters and it can be done and that is case by case.

dagar commented:
Another option to consider is to simply prevent the SPI transfer from an interrupt. Once https://github.com/PX4/Firmware/pull/11571 is merged the only concern is 3rd party drivers rebased on newer PX4. I think it would be acceptable to fail at runtime with an error (or even assert) if it points to instructions that show how trivial it is to migrate.
